# scikit-learn

## scikit-learnでできること

### 教師あり学習
- 回帰
- 分類

### 教師なし学習
- クラスタリング
- 次元削減
- データ前処理



In [227]:
from sklearn.datasets import load_boston

dataset = load_boston()

/Users/komotokenta/Docker/nuinfo-seminar/.nuinfo-seminar/lib/python3.10/site-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function load_boston is deprecated; `load_boston` is deprecated in 1.0 and will be removed in 1.2.

    The Boston housing prices dataset has an ethical problem. You can refer to
    the documentation of this function for further details.

    The scikit-learn maintainers therefore strongly discourage the use of this
    dataset unless the purpose of the code is to study and educate about
    ethical issues in data science and machine learning.

    In this special case, you can fetch the dataset from the original
    source::

        import pandas as pd
        import numpy as np

        data_url = "http://lib.stat.cmu.edu/datasets/boston"
        raw_df = pd.read_csv(data_url, sep="\s+", skiprows=22, header=None)
        data = np.hstack([raw_df.values[::2, :], raw_df.values[1::2, :2]])
        target = raw_df.values[1::2, 2]

    Alternative dataset

In [228]:
#データのキーの一覧を取得
#['data'(入力値. 今回の場合はMEDV以外の13個の値), 'target'(目標値. 今回の場合はMEDV), 'feature_names'(属性名), 'DESCR'(概要), 'filename'(ファイルの名前), 'data_module']
dataset.keys()

# dataset.key_name
# もしくは
# dataset['key_name']
# で値を取得できる

dict_keys(['data', 'target', 'feature_names', 'DESCR', 'filename', 'data_module'])

In [229]:
print(dataset.DESCR)

.. _boston_dataset:

Boston house prices dataset
---------------------------

**Data Set Characteristics:**  

    :Number of Instances: 506 

    :Number of Attributes: 13 numeric/categorical predictive. Median Value (attribute 14) is usually the target.

    :Attribute Information (in order):
        - CRIM     per capita crime rate by town
        - ZN       proportion of residential land zoned for lots over 25,000 sq.ft.
        - INDUS    proportion of non-retail business acres per town
        - CHAS     Charles River dummy variable (= 1 if tract bounds river; 0 otherwise)
        - NOX      nitric oxides concentration (parts per 10 million)
        - RM       average number of rooms per dwelling
        - AGE      proportion of owner-occupied units built prior to 1940
        - DIS      weighted distances to five Boston employment centres
        - RAD      index of accessibility to radial highways
        - TAX      full-value property-tax rate per $10,000
        - PTRATIO  pu

.. _boston_dataset:

Boston house prices dataset
---------------------------

**Data Set Characteristics:**  

    :Number of Instances: 506 

    :Number of Attributes: 13 numeric/categorical predictive. Median Value (attribute 14) is usually the target.

    :Attribute Information (in order):
        - CRIM     per capita crime rate by town
        - ZN       proportion of residential land zoned for lots over 25,000 sq.ft.
        - INDUS    proportion of non-retail business acres per town
        - CHAS     Charles River dummy variable (= 1 if tract bounds river; 0 otherwise)
        - NOX      nitric oxides concentration (parts per 10 million)
        - RM       average number of rooms per dwelling
        - AGE      proportion of owner-occupied units built prior to 1940
        - DIS      weighted distances to five Boston employment centres
        - RAD      index of accessibility to radial highways
        - TAX      full-value property-tax rate per $10,000
        - PTRATIO  pupil-teacher ratio by town
        - B        1000(Bk - 0.63)^2 where Bk is the proportion of black people by town
        - LSTAT    % lower status of the population
        - MEDV     Median value of owner-occupied homes in $1000's

    :Missing Attribute Values: None

    :Creator: Harrison, D. and Rubinfeld, D.L.

This is a copy of UCI ML housing dataset.
https://archive.ics.uci.edu/ml/machine-learning-databases/housing/


This dataset was taken from the StatLib library which is maintained at Carnegie Mellon University.

The Boston house-price data of Harrison, D. and Rubinfeld, D.L. 'Hedonic
prices and the demand for clean air', J. Environ. Economics & Management,
vol.5, 81-102, 1978.   Used in Belsley, Kuh & Welsch, 'Regression diagnostics
...', Wiley, 1980.   N.B. Various transformations are used in the table on
pages 244-261 of the latter.

The Boston house-price data has been used in many machine learning papers that address regression
problems.   
     
.. topic:: References

   - Belsley, Kuh & Welsch, 'Regression diagnostics: Identifying Influential Data and Sources of Collinearity', Wiley, 1980. 244-261.
   - Quinlan,R. (1993). Combining Instance-Based and Model-Based Learning. In Proceedings on the Tenth International Conference of Machine Learning, 236-243, University of Massachusetts, Amherst. Morgan Kaufmann.



### データセットの中身
| colmn | desc |
| - | :- |
| CRIM | 町ごとの一人当たり犯罪率 |
| ZN | 25,000sq.ft.以上の住宅用地に指定された土地の割合 |
| INDUS | 町ごとの非小売業用地面積の割合 |
| CHAS | チャールズ川のダミー変数（川に面している場合は1、そうでない場合は0）。 |
| NOX | 一酸化窒素濃度（1,000万分の1) |
| RM | 一戸あたりの平均部屋数 |
| AGE | 1940年以前に建てられた持ち家の割合 |
| DIS | ボストンの5つの雇用中心地までの距離を加重平均した |
| RAD | 放射状高速道路へのアクセス指数 |
| TAX | 1万ドルあたりの固定資産税評価額 |
| PTRATIO | 町ごとの生徒と教師の比率 |
| B | 1000(Bk - 0.63)^2 ただし、Bkは町ごとの黒人の割合である。 |
| LSTAT | 人口の下位地位の割合 |
| MEDV | 1000ドルの持ち家の中央値 |

In [230]:
type(dataset)

sklearn.utils._bunch.Bunch

## 回帰分析

In [231]:
x = dataset.data
t = dataset.target

print(x.shape)
print(t.shape)

(506, 13)
(506,)


### データセットを**訓練用**，**テスト用**に分ける

In [232]:
# データセットを分割する関数の読み込み
from sklearn.model_selection import train_test_split

# 訓練用データセットとテスト用データセットへの分割
x_train, x_test, t_train, t_test = train_test_split(x, t, test_size=0.3, random_state=0)
# test_size ... テスト用データセットの全体における比率（今回は30%をテスト用データセットに割り当てる）
# raodom_state ... この引数に毎回同じ数字を入れることで、実行のたびに値が変わることを防ぐことができる
print('x_train : ', x_train.shape)
print('x_test : ', x_test.shape)
print('t_train : ', t_train.shape)
print('t_test : ', t_test.shape)

x_train :  (354, 13)
x_test :  (152, 13)
t_train :  (354,)
t_test :  (152,)


### モデルの定義 & 学習

In [233]:
from sklearn.linear_model import LinearRegression

# モデルの定義
reg_model = LinearRegression()
# LinearRegression ... 重回帰分析を行うためのクラス. 最小二乗法を行うためのクラスで、目的関数や最適化手法も予め内部で用意されたものが使用される.
reg_model.fit(x_train, t_train)

LinearRegression()

精度の検証

In [234]:
print('parameter : \n', reg_model.coef_)             # 訓練後のパラメータ w
print('bias : ', reg_model.intercept_)               # 訓練後のバイアス b
print('score : ', reg_model.score(x_train, t_train)) # 精度の検証(決定係数)

parameter : 
 [-1.21310401e-01  4.44664254e-02  1.13416945e-02  2.51124642e+00
 -1.62312529e+01  3.85906801e+00 -9.98516565e-03 -1.50026956e+00
  2.42143466e-01 -1.10716124e-02 -1.01775264e+00  6.81446545e-03
 -4.86738066e-01]
bias :  37.93710774183255
score :  0.7645451026942549


In [235]:
x_test[:1]

array([[6.7240e-02, 0.0000e+00, 3.2400e+00, 0.0000e+00, 4.6000e-01,
        6.3330e+00, 1.7200e+01, 5.2146e+00, 4.0000e+00, 4.3000e+02,
        1.6900e+01, 3.7521e+02, 7.3400e+00]])

In [236]:
# 個別に推論する

print('predict : ', reg_model.predict(x_test[:1])) # 推論
print('target : ', t_test[0])

predict :  [24.9357079]
target :  22.6


In [237]:
print('score : ', reg_model.score(x_test, t_test)) # テスト用データセットでモデルを評価

score :  0.6733825506400184


↑ 過学習が起こっている？

### データを標準化してみる

In [238]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
# StandardScaler ... 標準化を行うためのクラス.
scaler.fit(x_train)

x_train_scaled = scaler.transform(x_train)
x_test_scaled  = scaler.transform(x_test)

reg_model = LinearRegression()
reg_model.fit(x_train_scaled, t_train)

print('score train : ', reg_model.score(x_train_scaled, t_train))
print('score test  : ', reg_model.score(x_test_scaled, t_test))

score train :  0.7645451026942549
score test  :  0.6733825506400196


↑ スコアは上がらなかった...

### べき変換してみる

In [239]:
from sklearn.preprocessing import PowerTransformer

scaler = PowerTransformer()
# PowerTransformer ... べき変換を行うためのクラス. べき変換というのは、データを正規分布に近くなるように正規化する変換のことで２種類（Box-CoxとYeo-Johnson）がある. 
#                      このクラスでの既定の方法はYeo-Johnson法
scaler.fit(x_train)
x_train_scaled = scaler.transform(x_train)
x_test_scaled  = scaler.transform(x_test)

reg_model = LinearRegression()

reg_model.fit(x_train_scaled, t_train)

print('score train : ', reg_model.score(x_train_scaled, t_train))
print('score test  : ', reg_model.score(x_test_scaled, t_test))

score train :  0.7958790042599264
score test  :  0.7135153164874446


In [240]:
print('predict : ', reg_model.predict(x_test_scaled[:1]))
print('target : ', t_test[0])

predict :  [24.82726017]
target :  22.6


### パイプライン化

In [241]:
# パイプライン化
from sklearn.pipeline import Pipeline

# パイプラインの作成 (scaler -> svr)
pipeline = Pipeline([
    ('scaler', PowerTransformer()),
    ('reg', LinearRegression())
])

# scaler および reg を順番に使用
pipeline.fit(x_train, t_train)

# 訓練用データセットを用いた決定係数の算出
print(pipeline.score(x_train, t_train))

# テスト用データセットを用いた決定係数の算出
linear_result = pipeline.score(x_test, t_test)

print(linear_result)

0.7958790042599264
0.7135153164874446


## 精度を上げてみよう！！

NNを使ってみる

In [242]:
from sklearn.neural_network import MLPRegressor

mlp_model = MLPRegressor(hidden_layer_sizes=(300,), max_iter=1000, random_state=0)

# LinearRegression ... 重回帰分析を行うためのクラス. 最小二乗法を行うためのクラスで、目的関数や最適化手法も予め内部で用意されたものが使用される.
mlp_model.fit(x_train, t_train)


MLPRegressor(hidden_layer_sizes=(300,), max_iter=1000, random_state=0)

In [243]:
print('train_score : ', mlp_model.score(x_train, t_train)) # 精度の検証(決定係数)
print('test_score  : ', mlp_model.score(x_test, t_test)) # 精度の検証(決定係数)

train_score :  0.8573368062874203
test_score  :  0.7256711901517957


In [244]:
# 個別に推論する
print('predict : ', mlp_model.predict(x_test[:1])) # 推論
print('target  : ', t_test[0])

predict :  [26.53777743]
target  :  22.6


In [245]:
from sklearn.neural_network import MLPRegressor

mlp_model2 = MLPRegressor(hidden_layer_sizes=(300,), max_iter=10000, random_state=0, solver="lbfgs")

# LinearRegression ... 重回帰分析を行うためのクラス. 最小二乗法を行うためのクラスで、目的関数や最適化手法も予め内部で用意されたものが使用される.
mlp_model2.fit(x_train, t_train)


/Users/komotokenta/Docker/nuinfo-seminar/.nuinfo-seminar/lib/python3.10/site-packages/sklearn/neural_network/_multilayer_perceptron.py:559: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
  self.n_iter_ = _check_optimize_result("lbfgs", opt_res, self.max_iter)


MLPRegressor(hidden_layer_sizes=(300,), max_iter=10000, random_state=0,
             solver='lbfgs')

In [247]:
print('train_score : ', mlp_model2.score(x_train, t_train)) # 精度の検証(決定係数)
print('test_score  : ', mlp_model2.score(x_test, t_test)) # 精度の検証(決定係数)

train_score :  0.9837600696370614
test_score  :  0.7870578666776937


In [248]:
from sklearn.neural_network import MLPRegressor

mlp_model2 = MLPRegressor(hidden_layer_sizes=(300,), max_iter=20000, random_state=0, solver="lbfgs")

mlp_model2.fit(x_train, t_train)

MLPRegressor(hidden_layer_sizes=(300,), max_iter=20000, random_state=0,
             solver='lbfgs')

In [249]:
print('train_score : ', mlp_model2.score(x_train, t_train)) # 精度の検証(決定係数)
print('test_score  : ', mlp_model2.score(x_test, t_test)) # 精度の検証(決定係数)

train_score :  0.984996885924859
test_score  :  0.7905471802792009
